ssl/ech/ech_internal.c

```diff
@@ -160,4 +217,940 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+    }
+    if (OSSL_HPKE_get_grease_value(hpke_suite_in_p, &hpke_suite,
+                                   senderpub, &senderpub_len,
+                                   cipher, cipher_len,
+                                   s->ssl.ctx->libctx, NULL) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (!WPACKET_put_bytes_u16(pkt, s->ext.ech.attempted_type)
+        || !WPACKET_start_sub_packet_u16(pkt)
+        || !WPACKET_put_bytes_u8(pkt, OSSL_ECH_OUTER_CH_TYPE)
+        || !WPACKET_put_bytes_u16(pkt, hpke_suite.kdf_id)
+        || !WPACKET_put_bytes_u16(pkt, hpke_suite.aead_id)
+        || !WPACKET_put_bytes_u8(pkt, cid)
+        || !WPACKET_sub_memcpy_u16(pkt, senderpub, senderpub_len)
+        || !WPACKET_sub_memcpy_u16(pkt, cipher, cipher_len)
+        || !WPACKET_close(pkt)
+        ) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /* record the ECH sent so we can re-tx same if we hit an HRR */
+    OPENSSL_free(s->ext.ech.sent);
+    WPACKET_get_total_written(pkt, &pp_at_end);
+    s->ext.ech.sent_len = pp_at_end - pp_at_start;
+    s->ext.ech.sent = OPENSSL_malloc(s->ext.ech.sent_len);
+    if (s->ext.ech.sent == NULL) {
+        s->ext.ech.sent_len = 0;
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    memcpy(s->ext.ech.sent, WPACKET_get_curr(pkt) - s->ext.ech.sent_len,
+           s->ext.ech.sent_len);
+    s->ext.ech.grease = OSSL_ECH_IS_GREASE;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "ECH - sending GREASE\n");
+    } OSSL_TRACE_END(TLS);
+    return 1;
+}
+
+/*
+ * Search the ECH store for one that's a match. If no outer_name was set via
+ * API then we just take the 1st match where we locally support the HPKE suite.
+ * If OTOH, an outer_name was provided via API then we prefer the first that
+ * matches that. Name comparison is via case-insensitive exact matches.
+ */
+int ossl_ech_pick_matching_cfg(SSL_CONNECTION *s, OSSL_ECHSTORE_ENTRY **ee,
+                               OSSL_HPKE_SUITE *suite)
+{
+    int namematch = 0, nameoverride = 0, suitematch = 0, num, cind = 0;
+    unsigned int csuite = 0, tsuite = 0, hnlen = 0;
+    OSSL_ECHSTORE_ENTRY *lee = NULL, *tee = NULL;
+    OSSL_ECHSTORE *es = NULL;
+    char *hn = NULL;
+
+    if (s == NULL || s->ext.ech.es == NULL || ee == NULL || suite == NULL)
+        return 0;
+    *ee = NULL;
+    es = s->ext.ech.es;
+    if (es->entries == NULL)
+        return 0;
+    num = sk_OSSL_ECHSTORE_ENTRY_num(es->entries);
+    /* allow API-set pref to override */
+    hn = s->ext.ech.outer_hostname;
+    hnlen = (hn == NULL ? 0 : strlen(hn));
+    if (hnlen != 0)
+        nameoverride = 1;
+    if (s->ext.ech.no_outer == 1) {
+        hn = NULL;
+        hnlen = 0;
+        nameoverride = 1;
+    }
+    for (cind = 0; cind != num && (suitematch == 0 || namematch == 0); cind++) {
```

Comment:
Nit: I prefer `<` instead of `!=` and it also handles the case when `es->entries == NULL` although you check that explicitly.